LogMonitor, Remove TranscodeVideoProfile, ClaimManager

ericxtang opened the pull request:
LogMonitor monitors the Ethereum log, and allows anyone to register a subscriber for events.

ericxtang committed:
add log manager

dob commented on eth/logmonitor.go:
```diff
@@ -0,0 +1,55 @@
+
+	go func() {
+		logsCh := make(chan types.Log)
+		logsSub, err := eth.SubscribeToJobEvent(context.Background(), logsCh)
```
Is this only useful for the `Job` event, rather than being generic to allow users to subscribe to all events?

ericxtang replied:
Hmm... I was hoping to move all log-related code in here.

dob commented on common/util.go:
```diff
@@ -0,0 +1,14 @@
+	start := time.Now()
+	for time.Since(start) < waitTime {
+		if condition() == false {
+			time.Sleep(100 * time.Millisecond)
```
Any reason in particular for using this `time.Sleep` pattern rather than the golang Timer pattern with channels? More just curious...not suggesting otherwise.

ericxtang replied:
I wanted to periodically check if a condition is true, and have a max total wait-time.  I think Timer only allows you to wait for the whole period right?

yondonfu commented on eth/logmonitor.go:
```diff
@@ -0,0 +1,55 @@
+	callbacks []func(j *Job)
+}
+
+func NewLogMonitor(eth LivepeerEthClient) *LogMonitor {
```
Isn't most of the logic in this file already in `jobmanager.go`?

ericxtang replied:
Right... I couldn't find anywhere that actually invokes the JobManager... Maybe it's not being used right now?  Agree we should probably combine the two.  Let's talk on Monday.

ericxtang committed:
pretty big change...

ericxtang commented:
Added a few things to this PR - Happy to do a in-person code review.
* Fixed claimManager for submitting proofs that combine all video profiles
* Removed TranscodeProfile, so now we JUST have VideoProfile
* Added Logmonitor - maybe we need to change this, but I have some questions about the current JobManager (maybe JobSubscription should to expose jidChan as a public field?)

ericxtang committed:
remove jobmanager

yondonfu commented on core/claimmanager.go:
```diff
@@ -131,161 +155,185 @@ func (c *BasicClaimManager) SufficientBroadcasterDeposit() (bool, error) {
 		}
 
-		glog.Infof("Submitting claim root: %v", root.Hash.Hex())
+		if startNewRange {
```
I think this can just be `if startNewRange && i+1 != len(keys)`

yondonfu commented on core/claimmanager.go:
```diff
@@ -131,161 +155,185 @@ func (c *BasicClaimManager) SufficientBroadcasterDeposit() (bool, error) {
+				}
+				glog.Infof("Got block hash: %x, block number: %v", blkHash, blkNum)
+				//Record claim information for verification later
+				for i := segRange[0]; i <= segRange[1]; i++ {
```
I think all the goroutines in this loop will close over the same variable `segRange`, but might not execute until the loop is finished resulting in the same value for `segRange` being used for each goroutine. Passing in `segRange` as a parameter to the goroutine so that `segRange` is evaluated at each iteration and the value is passed to the goroutine should fix it.

ericxtang replied:
Ahh - good call!  I wonder if I do this in other parts of the code base...

yondonfu commented on core/claimmanager.go:
```diff
@@ -93,25 +105,37 @@ func (c *BasicClaimManager) AddReceipt(seqNo int64, data []byte, tDataHash strin
+	if _, ok := cd.tDataHashes[profile]; ok {
+		return ErrClaimManager
 	}
+	cd.tDataHashes[profile] = tDataHash
```
I think the current assumption in the contracts is that each receipt contains the hash of all transcoded data for all video profiles while this function seems to add a receipt per video profile.

A thought: maybe a receipt should store an array of all the transcoded data hashes for all video profiles. The ClaimManager could keep track of the receipt struct and then when `Claim` is called, it would check if the receipt contains transcoded data hashes for all video profiles. If so, it could sort & concatenate the transcoded data hashes and then take the hash of the concatenation. The concatenation hash would then be used to generate the hash of the receipt.

ericxtang replied:
I think we are thinking about the same thing here.  claimConcatTDatahash gets populated with the "combined hash" when we call Claim().  We also check to make sure all the in makeRanges()

yondonfu commented on core/livepeernode.go:
```diff
@@ -138,10 +139,39 @@ func (n *LivepeerNode) EndTranscodeJob(jid *big.Int) error {
+func (n *LivepeerNode) ClaimVerifyAndDistributeFees(cm ClaimManager) error {
+	//Do the claim, wait until it's finished
+	count, rc, ec := cm.Claim()
+	select {
```
Are the contents of this select statement supposed to run once for each claim i.e. until count == 0? It looks like the select statement only runs once right now

ericxtang replied:
Yep you are right.  adding a test

yondonfu commented on core/livepeernode.go:
```diff
@@ -138,10 +139,39 @@ func (n *LivepeerNode) EndTranscodeJob(jid *big.Int) error {
+		}
+	}
+
+	if err := cm.Verify(); err != nil {
```
If there are multiple claims, there should be a call to `Verify()` and a call to `DistributeFees()` for each claim

ericxtang replied:
Verify and DistributeFees take care of that.  I think there would be unnecessary calls if Claim failed for some reason though...

yondonfu commented on core/livepeernode.go:
```diff
@@ -202,11 +232,17 @@ func (n *LivepeerNode) TranscodeAndBroadcast(config net.TranscodeConfig, cm Clai
 			}
 
+			if n.Eth != nil {
+				if jid := n.VideoDB.GetJidByStreamID(StreamID(config.StrmID)); jid != nil {
```
Does this mean a transcoder always calls `EndJob()` when a stream is finished?

ericxtang replied:
Added another check to make sure the job is not already ended.

yondonfu commented on eth/client.go:
```diff
@@ -524,7 +524,7 @@ func (c *Client) SubscribeToJobEvent(ctx context.Context, logsCh chan types.Log)
 	q := ethereum.FilterQuery{
 		Addresses: []common.Address{c.jobsManagerAddr},
-		Topics:    [][]common.Hash{[]common.Hash{abiJSON.Events["NewJob"].Id()}, []common.Hash{common.BytesToHash(common.LeftPadBytes(c.account.Address[:], 32))}},
+		Topics:    [][]common.Hash{[]common.Hash{abiJSON.Events["NewJob"].Id()}, []common.Hash{common.BytesToHash(common.LeftPadBytes(transcoderAddr[:], 32))}, []common.Hash{common.BytesToHash(common.LeftPadBytes(broadcasterAddr[:], 32))}},
```
According to these comments: https://github.com/ethereum/go-ethereum/blob/019dca9ba2f0ad91047129b95d91b95a9e71650e/interfaces.go an empty slice will match any topic. Looks like in `livepeer.go` we pass in an empty address if we don't care about the transcoder or broadcaster address for a job event, but `common.BytesToHash(common.LeftPadBytes(...))` might lead to a hash with all zeros? Not sure about this, but probably worth testing out and if that is what happens we can just explicitly pass in an empty slice for a topic we don't care about

ericxtang replied:
I tested it out with a local geth node - seems to work

yondonfu reviewed (commented):
Left some initial comments and will follow up with an in-person code review

ericxtang committed:
fixed according to commends

ericxtang committed:
in person code changes

ericxtang commented:
Did in-person code review
